### DISCUSSION TITLE
Is the compat dependency necessary?

### DISCUSSION THREAD

**Discussion by Fuco1** [emacscollective/no-littering#179] [General]
The `compat` dependency seems a bit broken, let me explain.

I'm using Emacs 26.3 and the latest version of that package *somehow* (I don't know how) defines function `json-parse-string` but *does not* also define the function `json-parse-buffer`.  

The `lsp-mode` package checks for existence of `json-parse-string` to use the native json parsing.  However, loading `no-littering` which loads `compat` breaks my Emacs environment, because suddenly only some functions from the `json-parse-*` family are available.  See also the upstream issue https://github.com/emacs-lsp/lsp-mode/issues/3536

I've went through the code of this package and I don't see why the `compat` dependency is necessary.  Should we consider dropping it?

**Comment by tarsius** [comment]
> I'm using Emacs 26.3 and the latest version of that package *somehow* (I don't know how) defines function `json-parse-string` but *does not* also define the function `json-parse-buffer`. 

If you haven't done so yet, you should report that to `compat`'s maintainer. Maybe he can make a change or maybe he can provide some guidance on how and why `lsp-mode` should adjust its check.

> I've went through the code of this package and I don't see why the `compat` dependency is necessary.  Should we consider dropping it?

Some of my most popular packages definitely need `compat` now. I figure that means that nearly all of the users of my less popular packages already have `compat` installed. I have decided to add `compat` as a dependency of all of my packages that I still have installed myself. Most of them really do need `compat`, a handful do not. I have done it anyway because then I can just use the functions that I know to be provided by `compat` without every time having to scroll to the beginning of the file to see whether `compat` is already required in that particular library/package.

It also makes it easier for other package maintainers to use `compat` if users already have it installed because I added it as a dependency. The "I don't want to install an additional, and in my case actually unnecessary, package just for your package" argument won't work anymore if some other package already "forced" them to install it. Though admittedly `magit` plays a much bigger role here than `no-littering`. `compat` makes the lifes of maintainers easier and sometimes doing something just for the benefit of maintainers is okay too (though the one user, who gave me a lot of shit when I dared to add `dash` as a dependency of `magit`, would probably disagree :shrug:).

**Reply by Fuco1** [reply]
Definitely agree with using dependencies and code reuse, that's the point.  I was just a little worried about the compat package looking a little bit broken.  I was going through the sources and honestly don't understand what it does and/or how :D

If this is in magit then I have no luck, because I would gladly exchange all other packages just to have magit :+1: 

I'm also trying to report this upstream, so we'll see.  Given the fact compat is included in magit, we can close this discussion here.
